pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,333 @@
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: configMapGVK, Key: "default/config-test-2"}: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+	// now assert that the gvkAggregator looks as expected
+	agg.IsPresent(configMapGVK)
+	gvks = agg.List(syncSourceOne)
+	require.Len(t, gvks, 1)
+	_, foundConfigMap = gvks[configMapGVK]
+	require.True(t, foundConfigMap)
+	_, foundPod = gvks[podGVK]
+	require.False(t, foundPod)
+
+	// now swap the gvks for each source and do so repeatedly to generate some churn
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		order := true
+		for i := 1; i <= 10; i++ {
+			if order {
+				require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+				require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceTwo, []schema.GroupVersionKind{podGVK}))
+			} else {
+				require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{podGVK}))
+				require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceTwo, []schema.GroupVersionKind{configMapGVK}))
+			}
+
+			order = !order
+		}
+
+		// final upsert for determinism
+		require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+		require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceTwo, []schema.GroupVersionKind{podGVK}))
+	}()
+
+	wg.Wait()
+
+	expected = map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: configMapGVK, Key: "default/config-test-2"}: nil,
+		{Gvk: podGVK, Key: "default/pod-1"}:               nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+	// now assert that the gvkAggregator looks as expected
+	agg.IsPresent(configMapGVK)
+	gvks = agg.List(syncSourceOne)
+	require.Len(t, gvks, 1)
+	_, foundConfigMap = gvks[configMapGVK]
+	require.True(t, foundConfigMap)
+	gvks = agg.List(syncSourceTwo)
+	require.Len(t, gvks, 1)
+	_, foundPod = gvks[podGVK]
+	require.True(t, foundPod)
+
+	// now remove the sources
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceOne))
+	}()
+	go func() {
+		defer wg.Done()
+		require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceTwo))
+	}()
+
+	wg.Wait()
+
+	// and expect an empty cache and empty aggregator
+	require.Eventually(t, expectedCheck(cfClient, map[cachemanager.CfDataKey]interface{}{}), eventuallyTimeout, eventuallyTicker)
+	require.True(t, len(agg.GVKs()) == 0)
+
+	// cleanup
+	require.NoError(t, c.Delete(ctx, cm), "deleting ConfigMap config-test-1")
+	require.NoError(t, c.Delete(ctx, cm2), "deleting ConfigMap config-test-2")
+	require.NoError(t, c.Delete(ctx, pod), "deleting Pod pod-1")
+}
+
+func expectedCheck(cfClient *cachemanager.FakeCfClient, expected map[cachemanager.CfDataKey]interface{}) func() bool {
```

Comment:
I wonder if it's possible to make this a receiver method on the `FakeCfClient`?  Then it would be easily re-usable.  The purpose of that fake client is for testing, so making it more useful in a test seems appropriate.

Reply:
chatted offline to track deferral in https://github.com/open-policy-agent/gatekeeper/issues/2929